fe/endpoint_correction.py

```diff
@@ -0,0 +1,199 @@
+import functools
+
+from typing import Tuple
+import jax
+import numpy as np
+import jax.numpy as jnp
+from timemachine.potentials import bonded
+from scipy.stats import special_ortho_group
+
+def exp_u(rotation, k, beta):
+    return jnp.exp(-beta*bonded.psi(rotation, k))
+
+exp_batch = jax.jit(jax.vmap(exp_u, (0, None, None)))
+
+def sample_multiple_rotations(k, beta, size):
+    num_batches = 500
+    batch_size = 10000
+
+    samples = []
+
+    for batch_attempt in range(num_batches):
+        Rs = special_ortho_group.rvs(3, size=batch_size)
+        tests = np.random.rand(batch_size)
+        M = np.pi**2 # volume of SO(3)
+        # acceptance_prob = exp_batch(Rs, k)/M
+        acceptance_prob = exp_batch(Rs, k, beta)
+        locations = np.argwhere(tests < acceptance_prob).reshape(-1)
+
+        samples.append(Rs[locations])
+        if sum([len(x) for x in samples]) > size:
+            break
+
+    return np.concatenate(samples)[:size]
+
+def estimate_delta_us(
+    k_translation,
+    k_rotation,
+    core_idxs,
+    core_params,
+    beta,
+    lhs_xs,
+    rhs_xs):
+    """
+    Compute the BAR re-weighted end-point correction of converting an intractable core
+    restraint into a tractable RMSD-based orientational restraint.
+
+    Parameters
+    ----------
+    k_translation: float
+        Force constant of the translational restraint
+
+    k_rotation: float
+        Force constant of the rotational restraint
+
+    core_idxs: int np.array (C, 2)
+        Atom mapping between the two cores
+
+    core_params: float np.array (C, 2)
+        Bonded parameters of the intractable restraint
+
+    lhs_xs: np.array [T, N, 3]
+        Samples from the intractable left hand state, with the restraints turned on.
+
+    rhs_xs: np.array [T, N, 3]
+        Samples from the non-interacting, fully unrestrained right hand state.
+
+    beta: 1/kT
+        inverse kT
+
+    Returns
+    -------
+    4-tuple
+        4-tuple array of lhs delta_us, rhs delta_us, sampled rotations, and sampled translations.
```

Comment:
I think this convention for lhs_du and rhs_du is now consistent with what @maxentile has in mind (and that of pymbar).